for #498: security onboarding panels and logic

groovecoder opened the pull request:
Not quite ready for merge yet. I need the image for the 3rd security onboarding panel.

To test:
* `shield run . control -- --binary Nightly` should show the existing onboarding panels
* `shield run . securityOnboarding -- --binary Nightly` should show the new security onboarding panels

groovecoder commented:
Adding @johngruen b/c the copy needs some love too.

jonathanKingston commented on webextension/js/popup.js:
```diff
@@ -69,15 +69,19 @@ const Logic = {
+    const variationPromise = this.getShieldStudyVariation();
 
     // Routing to the correct panel.
+    Promise.all([identitiesPromise, variationPromise])
```
`await Promise.all()` maybe?

jonathanKingston commented on webextension/js/popup.js:
```diff
@@ -257,6 +270,14 @@ const Logic = {
   },
 
+  getShieldStudyVariation() {
+    return browser.runtime.sendMessage({
```
await?

jonathanKingston commented on webextension/js/popup.js:
```diff
@@ -159,6 +163,15 @@ const Logic = {
     }).catch((e) => {throw e;});
   },
 
+  getPanelSelector(panel) {
```
This is fine, I perhaps would have used a translation based approach where the strings are embedded in some JSON and some method looks up the variation translation, this would potentially be a little less coupled to having this method.

groovecoder commented:
I cleaned up the first instance of `.then()` with `await Promise.all()`, but I stopped short of going into the rabbit-hole of `async`+`await` thru all of `refreshIdentities` and `getShieldStudyVariation`.

I think I started on a JSON/translation approach, but it was easier and simpler to check the variation in the single `showPanel` call?

jonathanKingston commented:
@groovecoder cool I'm happy for all of this just wanted to note that it won't really scale as a solution if we decided to spin of 5 variations etc.

jonathanKingston commented:
I'm getting `getShieldStudyVariation` returning undefined meaning both variations for me are the same.

Ah... `json` whatever command that is doesn't exist on Ubuntu.

jonathanKingston commented:
`jq .name=\"shield-study-containers\" package.json` does the same for `jq` which I would expect is more universal.

I could only get it working with this though (seems jq doesn't like writing to the same file):
```
jq .name=\"shield-study-containers\" package.json > shield.json && mv shield.json package.json
```

jonathanKingston commented:
Hey @groovecoder could you make the command the following:

```
    "package-shield": "./node_modules/json/lib/json.js -I -f package.json -e 'this.name=\"shield-study-containers\"' && jpm xpi && ./node_modules/json/lib/json.js -I -f package.json -e 'this.name=\"testpilot-containers\"'",
```

That way the dependency comes from the installed version rather than global. I didn't have the global version so it silently failed etc.

Also how would you generate the variations via `npm run`?

Thanks

groovecoder commented:
Added the 3rd onboarding panel image for the securityOnboarding variation.

Also, updated the command to use the local `json` binary.

@jonathanKingston - we don't generate different `.xpi` files to match the different variations. The shield library code sees the variations defined in the single `study.js` module and selects a variation for the user for the rest of the study's duration.

groovecoder commented:
Removing the DO NOT MERGE label, as I think we can merge this. But would still like to get the latest & greatest copy from @johngruen.

johngruen commented:
@groovecoder this has been formally reviewed: 

https://docs.google.com/document/d/1PgTjCvagjy3un64GfbIShvJBgi3FCe-YBd0XMdnm-Dc/edit?usp=sharing

groovecoder commented:
Updated to final copy. Ready for final r+ and merge @jonathanKingston

jonathanKingston commented:
@groovecoder I wasn't able to get this to work with test pilot for this branch in a brand new profile. It might just be me though. r+ and merge if the code works for you in this setup though and I can review again tonight.

groovecoder commented:
Ah, the security on-boarding panel only shows up when running the `@shield-study-containers` add-on.

jonathanKingston commented:
@groovecoder I mean I was expecting the normal onboarding. I didn't get this.

jonathanKingston commented:
Sorry I lied, it just doesn't even load the extension for me. No browser action or anything.

jonathanKingston commented:
I merged as I was able to reproduce in master which was more worrying,

jonathanKingston commented:
I was being a dork, new profiles don't allow unsigned xpi's. Sorry.
